src/HeapJobOrderer.ts

```diff
@@ -0,0 +1,156 @@
+import { Heap } from 'heap-js'
+import { Job } from './Job'
+import { JobOrderer } from './JobOrderer'
+import assert = require('assert')
+
+/**
+ * Manages a set of jobs that have to be run.
+ */
+export class HeapJobOrderer implements JobOrderer {
+	private sourcesHeap: Heap<Job>
+
+	private nonSources: Set<Job> = new Set()
+
+	private inProgress: Set<Job> = new Set()
+
+	private jobToDependents: Map<Job, Set<Job>> = new Map<Job, Set<Job>>()
+
+	private jobToCompletedPrereqs: Map<Job, Set<Job>> = new Map<Job, Set<Job>>()
+
+	/**
+	 * Create the JobOrderer and inform it of the jobs to manage.
+	 *
+	 * @throws When one of the jobs has a prerequisite that is not one of the jobs, or is itself.
+	 * @param jobs - The jobs to manage. They must have correctly populated prerequisites fields.
+	 */
+	constructor(rootJobs: Job[]) {
+		// Initialize the sources heap with its sorting function.
+		this.sourcesHeap = new Heap((job1, job2) => {
+			const job1Dependents = this.jobToDependents.get(job1)
+			const job2Dependents = this.jobToDependents.get(job2)
+
+			assert(
+				job1Dependents && job2Dependents,
+				'Provided jobs are missing from map.',
+			)
+
+			return job2Dependents.size - job1Dependents.size
+		})
+
+		// Add all of the root jobs to dependents map.
+		for (const rootJob of rootJobs) {
+			this.jobToDependents.set(rootJob, new Set())
+		}
+
+		const queue: Job[] = rootJobs
+		const seenSources: Set<Job> = new Set() // We use this because checking whether a heap has an element is O(n).
+
+		// BFS traverse dependency graph, starting with root nodes (the ultimate jobs).
+		while (queue.length) {
+			const job = queue.pop()
+			assert(job, 'Queue with nonzero length should have returned object.')
+
+			// Don't process the same node twice.
+			if (!seenSources.has(job) && !this.nonSources.has(job)) {
+				// Update job's prerequisites' 'dependents' fields.
+				for (const prerequisite of job.getPrerequisitesIterable()) {
+					let dependents = this.jobToDependents.get(prerequisite)
+
+					if (!dependents) {
+						dependents = new Set()
+						this.jobToDependents.set(prerequisite, dependents)
+					}
+
+					dependents.add(job)
+					queue.push(prerequisite)
```

Comment:
I might be missing something but why are we pushing the prerequisite back onto the queue. Wouldn't the queue already have all the jobs anyway? There is also the check for whether we have processed the same node above.

Reply:
We've seen `job` but not necessarily `job`'s prerequisites.